File: src/main/java/systems/conduit/main/api/factories/ItemFactory.java
package systems.conduit.main.api.factories;

import lombok.Builder;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.List;
import java.util.Map;

/**
 * Handles the creation of new itemstacks.
 *
 * @author Innectic
 * @since 11/28/2019
 */
@Builder
public class ItemFactory {

    private TextComponent name;
    private List<String> meta;
    private Item item;
    private Map<Enchantment, Integer> enchantments;
    private CompoundTag tag;
    private int count;
    private int repairCost;
    private int damageValue;

    public ItemStack item() {
        ItemStack stack = new ItemStack(item);

        if (name != null) stack.setHoverName(name);
        if (enchantments != null && !enchantments.isEmpty()) enchantments.forEach(stack::enchant);
        if (tag != null) stack.setTag(tag);
        if (count > 0) stack.setCount(count);
        if (repairCost > 0) stack.setRepairCost(repairCost);
        if (damageValue > 0) stack.setDamageValue(damageValue);

        return stack;
    }
}

File: src/main/kotlin/systems/conduit/api/factories/ItemFactory.kt
package systems.conduit.api.factories

import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.TextComponent
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.enchantment.Enchantment
import java.util.function.BiConsumer

/**
 * Handles the creation of new itemstacks.
 *
 * @author Innectic
 * @since 11/28/2019
 */
class ItemFactory {

    private val name: TextComponent? = null
    private val meta: List<String>? = null
    private val item: Item? = null
    private val enchantments: Map<Enchantment, Int>? = null
    private val tag: CompoundTag? = null
    private val count = 0
    private val repairCost = 0
    private val damageValue = 0

    fun item(): ItemStack {
        val stack = ItemStack(item)
        if (name != null) stack.hoverName = name
        if (enchantments != null && !enchantments.isEmpty()) enchantments.forEach(BiConsumer { enchantment: Enchantment?, i: Int? -> stack.enchant(enchantment, i!!) })
        if (tag != null) stack.tag = tag
        if (count > 0) stack.count = count
        if (repairCost > 0) stack.setRepairCost(repairCost)
        if (damageValue > 0) stack.damageValue = damageValue
        return stack
    }
}
